jcl/src/java.base/share/classes/java/lang/invoke/CollectHandle.java

```diff
@@ -35,25 +36,66 @@
  * down the call chain.
  */
 final class CollectHandle extends MethodHandle {
+	static final boolean[] EMPTY_BOOLEAN_ARRAY = new boolean[0];
+	static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+	static final char[] EMPTY_CHAR_ARRAY = new char[0];
+	static final short[] EMPTY_SHORT_ARRAY = new short[0];
+	static final int[] EMPTY_INT_ARRAY = new int[0];
+	static final long[] EMPTY_LONG_ARRAY = new long[0];
+	static final float[] EMPTY_FLOAT_ARRAY = new float[0];
+	static final double[] EMPTY_DOUBLE_ARRAY = new double[0];
 	@VMCONSTANTPOOL_FIELD
 	final MethodHandle next;
 	@VMCONSTANTPOOL_FIELD
 	final int collectArraySize; /* Size of the collect array */
 	@VMCONSTANTPOOL_FIELD
 	final int collectPosition; /* The starting position of arguments to collect */
+	final Object emptyArray;
 	
 	CollectHandle(MethodHandle next, int collectArraySize, int collectPosition) {
 		super(collectMethodType(next.type(), collectArraySize, collectPosition), KIND_COLLECT, new int[]{collectArraySize, collectPosition});
 		this.collectPosition = collectPosition;
 		this.collectArraySize = collectArraySize;
 		this.next = next;
+		if (collectArraySize == 0) {
+			emptyArray = createEmptyArray(next.type.arguments[collectPosition].getComponentType());
+		} else {
+			emptyArray = null;
+		}
 	}
 	
 	CollectHandle(CollectHandle original, MethodType newType) {
 		super(original, newType);
 		this.collectPosition = original.collectPosition;
 		this.collectArraySize = original.collectArraySize;
 		this.next = original.next;
+		this.emptyArray = original.emptyArray;
+	}
+
+	private static final Object createEmptyArray(Class<?> componentType) {
+		Object emptyArray;
+
+		if (componentType == boolean.class) {
```

Comment:
Does the benefit of doing all these type checks and deduplicating the empty primitive arrays have a measurable benefit in JRuby/Nashorn (since the object arrays aren't deduplicated)?

Reply:
@andrewcraik There is no measurable benefit in JRuby/Nashorn. The object arrays aren't deduplicated because the allocation of an object array may throw exception, so we can't pre-allocate it.